arrow-libs/fx/arrow-fx-resilience/src/commonMain/kotlin/arrow/fx/resilience/CircuitBreaker.kt

```diff
@@ -212,7 +212,13 @@ private constructor(
       is Closed -> {
         when (result) {
           is Either.Right -> {
-            result.value
+            val openingStrategy = state.get().openingStrategy
+            if (openingStrategy is OpeningStrategy.Count && openingStrategy.failuresCount == 0)
+              result.value
+            else {
+              state.compareAndSet(curr, Closed(openingStrategy.resetFailuresCount()))
+              result.value
```

Comment:
This is not a correct atomic update, `compareAndSet` returns `false` a parallel call already updated the `atomic` in-between. You need to _loop_ until `compareAndSet` returns `true`, the pattern in this data type is to use _tailrec recursion_ to do so. Alternatively you could also use `atomic.loop { }` with `return`.

The original code also used something similar to this:

```suggestion
              if(!state.compareAndSet(curr, Closed(openingStrategy.resetFailuresCount()))) markOrResetFailures(reset) // retry
              else result.value
```

Reply:
Good catch. It was a last-minute change so I totally forgot to add the retry piece